File: project_euler/src/problem_102.erl
%% @author std-string

%% Three distinct points are plotted at random on a Cartesian plane, for which -1000 ≤ x, y ≤ 1000, such that a triangle is formed.
%% Consider the following two triangles:
%% A(-340,495), B(-153,-910), C(835,-947)
%% X(-175,41), Y(-421,-714), Z(574,-645)
%% It can be verified that triangle ABC contains the origin, whereas triangle XYZ does not.
%% Using problem_012.dat, a data file containing the co-ordinates of one thousand "random" triangles, find the number of triangles for which the interior contains the origin.
%% NOTE: The first two examples in the file represent the triangles in the example given above.

-module(problem_102).
-export([get_check_data/0, prepare_data/2, solve/1]).

-behaviour(numerical_task_behaviour).

%% ====================================================================
%% API functions
%% ====================================================================

-spec get_check_data() -> [{Input :: term(), Output :: term()}].
get_check_data() -> [{"problem_102.dat", 228}].

-spec prepare_data(ModuleSourceDir :: string(), Input :: term()) -> term().
prepare_data(ModuleSourceDir, Filename) ->
    load_utils:read_number_table(filename:join(ModuleSourceDir, Filename), ",").

-spec solve(PreparedInput :: term()) -> term().
solve(VertexTable) ->
    VertexTable,
    CollectFun = fun(Row, Count) ->
        case check_triangle(Row) of
            true -> Count + 1;
            false -> Count
        end
    end,
    lists:foldl(CollectFun, 0, VertexTable).

%% ====================================================================
%% Internal functions
%% ====================================================================

-spec check_triangle(Triangle :: [integer()]) -> boolean().
check_triangle([X1, Y1, X2, Y2, X3, Y3]) ->
    ProductZ1 = claclulate_vectorz_product(X1, Y1, X2, Y2),
    ProductZ2 = claclulate_vectorz_product(X2, Y2, X3, Y3),
    ProductZ3 = claclulate_vectorz_product(X3, Y3, X1, Y1),
    check_vectorz_products(ProductZ1, ProductZ2, ProductZ3).

-spec claclulate_vectorz_product(X1 :: integer(), Y1 :: integer(), X2 :: integer(), Y2 :: integer()) -> integer().
claclulate_vectorz_product(X1, Y1, X2, Y2) -> Y1 * X2 - X1 * Y2.

-spec check_vectorz_products(ProductZ1 :: integer(), ProductZ2 :: integer(), ProductZ3 :: integer()) -> boolean().
check_vectorz_products(ProductZ1, ProductZ2, ProductZ3) when ProductZ1 > 0, ProductZ2 > 0, ProductZ3 > 0 -> true;
check_vectorz_products(ProductZ1, ProductZ2, ProductZ3) when ProductZ1 < 0, ProductZ2 < 0, ProductZ3 < 0 -> true;
check_vectorz_products(_ProductZ1, _ProductZ2, _ProductZ3) -> false.

File: projecteuler.fsharp/ProjectEulerTasks/Problem102.fs
namespace ProjectEulerTasks

open NUnit.Framework
open ProjectEulerTasks.Utils
open System.IO

// Three distinct points are plotted at random on a Cartesian plane, for which -1000 <= x, y <= 1000, such that a triangle is formed.
// Consider the following two triangles:
// A(-340,495), B(-153,-910), C(835,-947)
// X(-175,41), Y(-421,-714), Z(574,-645)
// It can be verified that triangle ABC contains the origin, whereas triangle XYZ does not.
// Using problem_012.dat, a data file containing the co-ordinates of one thousand "random" triangles, find the number of triangles for which the interior contains the origin.
// NOTE: The first two examples in the file represent the triangles in the example given above.

// TODO (std_string) : probably, move into CommonLib
module Problem102Impl =
    type Point = {X: int; Y: int}

open Problem102Impl

[<TestFixture>]
type Problem102() =

    let calcVectorZProduct (point1: Point) (point2: Point) = point1.Y * point2.X - point1.X * point2.Y

    let checkTriangle([point1; point2; point3]: Point list) =
        let productZ1 = calcVectorZProduct point1 point2
        let productZ2 = calcVectorZProduct point2 point3
        let productZ3 = calcVectorZProduct point3 point1
        match productZ1, productZ2, productZ3 with
        | _ when (productZ1 > 0) && (productZ2 > 0) && (productZ3 > 0) -> true
        | _ when (productZ1 < 0) && (productZ2 < 0) && (productZ3 < 0) -> true
        | _ -> false

    let solveImpl (dataFilename: string) =
        let createTriangle (numbers: int[]) =
            [{Point.X = numbers.[0]; Point.Y = numbers.[1]}; {Point.X = numbers.[2]; Point.Y = numbers.[3]}; {Point.X = numbers.[4]; Point.Y = numbers.[5]}]
        File.ReadAllLines(Path.Combine("Data", dataFilename)) |>
        Seq.map (fun line -> line.Split(',') |> Seq.map int |> Seq.toArray |> createTriangle) |>
        Seq.filter (fun points -> points |> checkTriangle) |>
        Seq.length

    [<TestCase("problem_102.dat", 228, TimeThresholds.HardTimeLimit)>]
    member public this.Solve(dataFilename: string, expectedAnswer: int, timeLimit: int) =
        SolutionUtils.CheckSolution(timeLimit, expectedAnswer, solveImpl, dataFilename)